### ISSUE TITLE
[Stripe] Subscribing to same plan twice results in two subscriptions

### ISSUE THREAD

**Issue by carter2422** [restrictcontentpro/restrict-content-pro#453]
If you have an active subscription and then *upgrade* to the same subscription you're then granted two subscriptions. 

![screen shot 2016-04-14 at 3 18 29 pm](https://cloud.githubusercontent.com/assets/1648631/14542414/6e18f194-0254-11e6-85ca-86fd20855a04.png)

Unless I'm mistaken the subscription terms should simply be modified (or replaced) rather than a second subscription being added. Is this correct?

**Comment by carter2422** [comment]
It seems that the core issue may be that the registration page is not always correctly filtering out the users current subscription level?  On some users I see only other levels as upgrade options on the registration page, but in the example above I was able to upgrade to the same subscription I was currently subscribed to.

**Comment by carter2422** [comment]
Additional evidence for the user shown above. 

The user was subscribed twice in Stripe, and invoiced for each subscription, but RCP treated it as proration of sorts: 
![screen shot 2016-04-14 at 3 24 55 pm](https://cloud.githubusercontent.com/assets/1648631/14542547/10aefc78-0255-11e6-8364-c1433942df57.png)
![screen shot 2016-04-14 at 3 26 04 pm](https://cloud.githubusercontent.com/assets/1648631/14542581/3d220502-0255-11e6-9277-52483af0e708.png)

**Comment by pippinsplugins** [comment]
Was the payment for the first subscription recorded in RCP?

**Comment by carter2422** [comment]
@pippinsplugins yes, as was the second subscription's payment (for $0).

**Comment by pippinsplugins** [comment]
How did they upgrade to the same subscription? Is the subscription level that the user is currently subscribed to still showing in the list (it shouldn't)?

**Comment by carter2422** [comment]
I'm not sure @pippinsplugins; I can't consistently replicate it, but for the test user above their current subscription was not filtered out of the list on the registration page. I simply chose the same subscription I was already subscribed to and *upgraded* again. 

In most cases the current subscription level is not an option, so I'm guessing there's a bug causing the filter to fail from time to time. In these scenarios it's probably worth putting in some further validation on the gateway to ensure you can't resubscribe to the same level.

**Comment by pippinsplugins** [comment]
The screenshot above looks like they purchased and then upgraded in the span of less than an hour. That makes me wonder if the Stripe webhook (which did have a system delay in the last few days) simply hadn't processed yet so the account wasn't marked as active. That would (I think) result in this kind of behavior.

**Comment by carter2422** [comment]
The webhook definitely worked as my test user was active on the first subscription prior to upgrading.

**Comment by pippinsplugins** [comment]
Were the screenshots above from your test or the real customer?

**Comment by carter2422** [comment]
They were from me signing up with a brand new, test user.

**Comment by mindctrl** [comment]
@carter2422 did you happen to be logged out?

**Comment by carter2422** [comment]
@mindctrl nope, was logged in the whole time.

**Comment by mindctrl** [comment]
Thanks @carter2422. 

We got another report of this yesterday from someone using Stripe. He thought it was related to the card update form, which isn't the case. He's now wondering how his customer triggered the bug that caused him to have two subscriptions of the same level. His second subscription appeared to be prorated as well.

**Comment by carter2422** [comment]
I've not yet updated to 2.5 on the live site, but I'll continue testing on dev to see if I can replicate this again @mindctrl

**Comment by pippinsplugins** [comment]
This might be resolved with https://github.com/restrictcontentpro/restrict-content-pro/pull/477 but need to test to confirm.

**Comment by pippinsplugins** [comment]
I can't get the level I currently have to ever show up on the form.

**Comment by pippinsplugins** [comment]
Aha, maybe I just tracked it down.

When an account is registered, the subscription that the member subscribes to is stored in a pending meta row. Once the webhook is processed from Stripe to confirm the subscription, the meta row is updated and the subscription is moved from the pending slot to the active slot. If the member, however, tries to subscribe again before the webhook is processed (or if the webhook fails to be processed), the same subscription level they just subscribed to will be shown on the registration form.

I don't think this is a bug. It's simply the nature of waiting for account verification.

The issue with not seeing the previous level cancelled, however, is definitely a bug, but that has been fixed in #477.

I feel we can mark this as fixed.
